In [1]:
# Import Libraries
import pandas as pd
import requests as rq
import numpy as np
from bs4 import BeautifulSoup
from sklearn.cluster import KMeans
import folium
from matplotlib import pyplot as plt
from matplotlib import cm, colors

# Part 1 : Creating a dataframe

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wikiResponse = rq.get(url)
soup = BeautifulSoup(wikiResponse.text,'lxml') 

items = []
rows = soup.table.find_all('tr')
for row in rows[1:]:
    cols = row.find_all('td')
    postal_code = cols[0].text
    borough = cols[1].text
    neighborhood = cols[2].text.strip()
    items.append([postal_code, borough, neighborhood])

In [3]:
# The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
df = pd.DataFrame(items, columns=["PostalCode", "Borough", "Neighborhood"])

df.head(10)

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront
5        M5A  Downtown Toronto       Regent Park
6        M6A        North York  Lawrence Heights
7        M6A        North York    Lawrence Manor
8        M7A      Queen's Park      Not assigned
9        M8A      Not assigned      Not assigned

In [4]:
# Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
df.drop(df[df['Borough']=='Not assigned'].index, inplace=True) # Drop inplace
df.reset_index(inplace=True, drop=True)

df.head(10)

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights
5        M6A        North York    Lawrence Manor
6        M7A      Queen's Park      Not assigned
7        M9A         Etobicoke  Islington Avenue
8        M1B       Scarborough             Rouge
9        M1B       Scarborough           Malvern

"Not assigned" Boroughs are now removed

In [5]:
# If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. 
# So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will 
# be Queen's Park.

df['Neighborhood'].where(df['Neighborhood']!='Not assigned', df['Borough'], inplace=True)
df.head(10)



PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights
5        M6A        North York    Lawrence Manor
6        M7A      Queen's Park      Queen's Park
7        M9A         Etobicoke  Islington Avenue
8        M1B       Scarborough             Rouge
9        M1B       Scarborough           Malvern

As you can see, Queen's Park Borough has the neighborhood same as the borough (index 6)

In [6]:
# More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, 
# you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. 
# These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 
# in the above table.

df = df.groupby(by=["PostalCode", "Borough"], as_index=False).aggregate(lambda neighborhoods: ", ".join(set(neighborhoods)))
df.reset_index(inplace=True, drop=True)

df.head(20)


PostalCode      Borough                                       Neighborhood
0         M1B  Scarborough                                     Rouge, Malvern
1         M1C  Scarborough             Rouge Hill, Highland Creek, Port Union
2         M1E  Scarborough                  Morningside, West Hill, Guildwood
3         M1G  Scarborough                                             Woburn
4         M1H  Scarborough                                          Cedarbrae
5         M1J  Scarborough                                Scarborough Village
6         M1K  Scarborough        Ionview, Kennedy Park, East Birchmount Park
7         M1L  Scarborough                    Oakridge, Clairlea, Golden Mile
8         M1M  Scarborough    Scarborough Village West, Cliffcrest, Cliffside
9         M1N  Scarborough                        Birch Cliff, Cliffside West
10        M1P  Scarborough  Wexford Heights, Scarborough Town Centre, Dors...
11        M1R  Scarborough                                  Maryvale, Wexford
12        M1S  Scarborough                                          Agincourt
13        M1T  Scarborough            Clarks Corners, Sullivan, Tam O'Shanter
14        M1V  Scarborough  L'Amoreaux East, Agincourt North, Steeles East...
15        M1W  Scarborough                                    L'Amoreaux West
16        M1X  Scarborough                                        Upper Rouge
17        M2H   North York                                  Hillcrest Village
18        M2J   North York                       Fairview, Henry Farm, Oriole
19        M2K   North York                                    Bayview Village

In [7]:
# In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.
df.shape

(103, 3)

There are 103 unique postalcode, borough pairs

# Part 2 : Use the Geocoder package or the csv file to create the dataframe



In [8]:
df_coord = pd.read_csv('https://cocl.us/Geospatial_data')
# Match the column names (Postal Code -> Post)
df_coord.rename(index=str, columns={"Postal Code":"PostalCode"},inplace=True)

df_coord.head(10)

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476
5        M1J  43.744734 -79.239476
6        M1K  43.727929 -79.262029
7        M1L  43.711112 -79.284577
8        M1M  43.716316 -79.239476
9        M1N  43.692657 -79.264848

In [9]:
df_geo = df.merge(df_coord, how='inner', on='PostalCode')

df_geo.head(10)


PostalCode      Borough                                     Neighborhood  \
0        M1B  Scarborough                                   Rouge, Malvern   
1        M1C  Scarborough           Rouge Hill, Highland Creek, Port Union   
2        M1E  Scarborough                Morningside, West Hill, Guildwood   
3        M1G  Scarborough                                           Woburn   
4        M1H  Scarborough                                        Cedarbrae   
5        M1J  Scarborough                              Scarborough Village   
6        M1K  Scarborough      Ionview, Kennedy Park, East Birchmount Park   
7        M1L  Scarborough                  Oakridge, Clairlea, Golden Mile   
8        M1M  Scarborough  Scarborough Village West, Cliffcrest, Cliffside   
9        M1N  Scarborough                      Birch Cliff, Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

# Part 3 : Explore and cluster the neighborhoods in Toronto.

In [10]:
# CLIENT_ID = 
# CLIENT_SECRET = 
# VERSION = '20180605'
# LIMIT = 100

In [11]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
#         print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = rq.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [12]:
toronto_venues = getNearbyVenues(names=df_geo['Neighborhood'],
                                   latitudes=df_geo['Latitude'],
                                   longitudes=df_geo['Longitude'])



In [13]:
toronto_venues.head(10)

Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1                          Rouge, Malvern              43.806686   
2  Rouge Hill, Highland Creek, Port Union              43.784535   
3  Rouge Hill, Highland Creek, Port Union              43.784535   
4       Morningside, West Hill, Guildwood              43.763573   
5       Morningside, West Hill, Guildwood              43.763573   
6       Morningside, West Hill, Guildwood              43.763573   
7       Morningside, West Hill, Guildwood              43.763573   
8       Morningside, West Hill, Guildwood              43.763573   
9       Morningside, West Hill, Guildwood              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.194353            Interprovincial Group       43.805630   
2              -79.160497           Chris Effects Painting       43.784343   
3              -79.160497            Royal Canadian Legion       43.782533   
4              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
5              -79.188711                G & G Electronics       43.765309   
6              -79.188711                       Marina Spa       43.766000   
7              -79.188711                 Big Bite Burrito       43.766299   
8              -79.188711                     chatr Mobile       43.765917   
9              -79.188711            Enterprise Rent-A-Car       43.764076   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.200378                  Print Shop  
2       -79.163742  Construction & Landscaping  
3       -79.163085                         Bar  
4       -79.189914                 Pizza Place  
5       -79.191537           Electronics Store  
6       -79.191000                         Spa  
7       -79.190720          Mexican Restaurant  
8       -79.191672                Tech Startup  
9       -79.193406         Rental Car Location

Let's check the size of the resulting dataframe



In [14]:
print(toronto_venues.shape)
toronto_venues.head()

(2258, 7)


Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1                          Rouge, Malvern              43.806686   
2  Rouge Hill, Highland Creek, Port Union              43.784535   
3  Rouge Hill, Highland Creek, Port Union              43.784535   
4       Morningside, West Hill, Guildwood              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.194353            Interprovincial Group       43.805630   
2              -79.160497           Chris Effects Painting       43.784343   
3              -79.160497            Royal Canadian Legion       43.782533   
4              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.200378                  Print Shop  
2       -79.163742  Construction & Landscaping  
3       -79.163085                         Bar  
4       -79.189914                 Pizza Place

Let's check how many venues were returned for each neighborhood


In [15]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Albion Gardens, Thistletown, Jamestown, Humberg...                     11   
Bathurst Quay, South Niagara, CN Tower, King an...                     16   
Bayview Village                                                         4   
Berczy Park                                                            55   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           22   
Business Reply Mail Processing Centre 969 Eastern                      19   
CFB Toronto, Downsview East                                             3   
Cabbagetown, St. James Town                                            46   
Caledonia-Fairbanks                                                     6   
Canada Post Gateway Processing Centre                                  10   
Cedarbrae                                                               8   
Central Bay Street                                                     88   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               15   
Church and Wellesley                                                   87   
Clarks Corners, Sullivan, Tam O'Shanter                                10   
Davisville                                                             38   
Davisville North                                                        8   
Del Ray, Mount Dennis, Keelesdale, Silverthorn                          4   
Design Exchange, Toronto Dominion Centre                              100   
Don Mills North                                                         5   
Dovercourt Village, Dufferin                                           20   
Downsview Central                                                       4   
Downsview Northwest                                                     5   
Downsview West                                                          3   
Downsview, Upwood Park, North Park                                      5   
East Toronto                                                            3   
Emery, Humberlea                                                        1   
...                                                                   ...   
Runnymede, Swansea                                                     38   
Ryerson, Garden District                                              100   
Scarborough Village                                                     1   
Scarborough Village West, Cliffcrest, Cliffside                         3   
Silver Hills, York Mills                                                1   
St. James Town                                                        100   
St. Phillips, Richview Gardens, Kingsview Villa...                      4   
Stn A PO Boxes 25 The Esplanade                                        94   
Studio District                                                        38   
The Beaches                                                             5   
The Beaches West, India Bazaar                                         18   
The Junction North, Runnymede                                           4   
The Junction South, High Park                                          23   
The Kingsway, Montgomery Road, Old Mill North                           3   
Thorncliffe Park                                                       18   
Trinity, Little Portugal                                               62   
Underground city, First Canadian Place                                100   
University of Toronto, Harbord                                         34   
Victoria Hotel, Commerce Court                                        100   
Victoria Village                                

Let's find out how many unique categories can be curated from all the returned venues


In [16]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 280 uniques categories.


## Analyze each neighborhood

In [17]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  ...  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0  ...      0              0                              0   
1                 0  ...      0              0                              0   
2                 0  ...      0              0                              0   
3                 0  ...      0              0                              0   
4                 0  ...      0              0                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wings Joint  Women's Store  
0         0            0              0  
1         0            0              0  
2         0            0              0  
3         0            0              0  
4         0            0              0  

[5 rows x 280 columns]

In [18]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                          Agincourt          0.0   
1  Albion Gardens, Thistletown, Jamestown, Humber...          0.0   
2  Bathurst Quay, South Niagara, CN Tower, King a...          0.0   
3                                    Bayview Village          0.0   
4                                        Berczy Park          0.0   

   Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0                0.0             0.0                0.0   0.0000   
1                0.0             0.0                0.0   0.0000   
2                0.0             0.0                0.0   0.0625   
3                0.0             0.0                0.0   0.0000   
4                0.0             0.0                0.0   0.0000   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  ...  \
0              0.0000        0.0000           0.000            0.000  ...   
1              0.0000        0.0000           0.000            0.000  ...   
2              0.0625        0.0625           0.125            0.125  ...   
3              0.0000        0.0000           0.000            0.000  ...   
4              0.0000        0.0000           0.000            0.000  ...   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0    0.0            0.0                       0.000000               0.0   
1    0.0            0.0                       0.000000               0.0   
2    0.0            0.0                       0.000000               0.0   
3    0.0            0.0                       0.000000               0.0   
4    0.0            0.0                       0.018182               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0     0.000000                    0.0              0.0       0.0          0.0   
1     0.090909                    0.0              0.0       0.0          0.0   
2     0.000000                    0.0              0.0       0.0          0.0   
3     0.000000                    0.0              0.0       0.0          0.0   
4     0.000000                    0.0              0.0       0.0          0.0   

   Women's Store  
0            0.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0  

[5 rows x 280 columns]

In [19]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [20]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                          Agincourt                Lounge   
1  Albion Gardens, Thistletown, Jamestown, Humber...         Grocery Store   
2  Bathurst Quay, South Niagara, CN Tower, King a...       Airport Service   
3                                    Bayview Village    Chinese Restaurant   
4                                        Berczy Park           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot        Sandwich Place    Chinese Restaurant   
1            Beer Store   Fried Chicken Joint  Fast Food Restaurant   
2      Airport Terminal        Airport Lounge       Harbor / Marina   
3                  Café   Japanese Restaurant                  Bank   
4          Cocktail Bar                Bakery    Seafood Restaurant   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant   Empanada Restaurant   
1          Liquor Store                     Pharmacy           Pizza Place   
2           Coffee Shop                        Plane      Sculpture Garden   
3         Women's Store                        Diner        Discount Store   
4    Italian Restaurant                   Steakhouse                  Café   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Dumpling Restaurant             Drugstore       Department Store  
1        Sandwich Place           Coffee Shop            Video Store  
2              Boutique                   Bar          Boat or Ferry  
3               Dog Run      Doner Restaurant             Donut Shop  
4           Cheese Shop              Beer Bar         Farmers Market

## Clustering neighborhoods

In [21]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 1, 0], dtype=int32)

In [22]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_geo

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

# Drop NaN
toronto_merged.dropna(axis=0, how='any',inplace=True)
toronto_merged.reset_index(inplace=True, drop=True)

In [23]:
toronto_merged.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Rouge Hill, Highland Creek, Port Union  43.784535   
2        M1E  Scarborough       Morningside, West Hill, Guildwood  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue  \
0 -79.194353             0.0  Fast Food Restaurant   
1 -79.160497             0.0                   Bar   
2 -79.188711             0.0     Electronics Store   
3 -79.216917             0.0           Coffee Shop   
4 -79.239476             0.0      Hakka Restaurant   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Print Shop         Deli / Bodega          Dessert Shop   
1  Construction & Landscaping         Women's Store    Dim Sum Restaurant   
2              Breakfast Spot    Mexican Restaurant          Tech Startup   
3           Korean Restaurant     Convenience Store         Women's Store   
4             Thai Restaurant                  Bank                Bakery   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Dim Sum Restaurant                 Diner        Discount Store   
1                 Diner        Discount Store               Dog Run   
2          Intersection        Medical Center   Rental Car Location   
3             Drugstore                 Diner        Discount Store   
4   Fried Chicken Joint  Caribbean Restaurant    Athletics & Sports   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dog Run      Doner Restaurant             Donut Shop  
1      Doner Restaurant            Donut Shop              Drugstore  
2           Pizza Place                   Spa             Donut Shop  
3               Dog Run      Doner Restaurant             Donut Shop  
4                Lounge          Costume Shop     Falafel Restaurant

In [24]:
# Let's check how many neighborhoods belong to each cluster
toronto_merged.groupby('Cluster Labels').count()

PostalCode  Borough  Neighborhood  Latitude  Longitude  \
Cluster Labels                                                           
0.0                     82       82            82        82         82   
1.0                     13       13            13        13         13   
2.0                      1        1             1         1          1   
3.0                      2        2             2         2          2   
4.0                      2        2             2         2          2   

                1st Most Common Venue  2nd Most Common Venue  \
Cluster Labels                                                 
0.0                                82                     82   
1.0                                13                     13   
2.0                                 1                      1   
3.0                                 2                      2   
4.0                                 2                      2   

                3rd Most Common Venue  4th Most Common Venue  \
Cluster Labels                                                 
0.0                                82                     82   
1.0                                13                     13   
2.0                                 1                      1   
3.0                                 2                      2   
4.0                                 2                      2   

                5th Most Common Venue  6th Most Common Venue  \
Cluster Labels                                                 
0.0                                82                     82   
1.0                                13                     13   
2.0                                 1                      1   
3.0                                 2                      2   
4.0                                 2                      2   

                7th Most Common Venue  8th Most Common Venue  \
Cluster Labels                                                 
0.0                                82                     82   
1.0                                13                     13   
2.0                                 1                      1   
3.0                                 2                      2   
4.0                                 2                      2   

                9th Most Common Venue  10th Most Common Venue  
Cluster Labels                                                 
0.0                                82                      82  
1.0                                13                      13  
2.0                                 1                       1  
3.0                                 2                       2  
4.0                                 2                       2

In [25]:
# create map
map_clusters = folium.Map(location=[df_geo['Latitude'].mean(), df_geo['Longitude'].mean()], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        # has to cast int to 'cluster' since it has a type float
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Cluster 0 

In [26]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood  \
0                                      Rouge, Malvern   
1              Rouge Hill, Highland Creek, Port Union   
2                   Morningside, West Hill, Guildwood   
3                                              Woburn   
4                                           Cedarbrae   
6         Ionview, Kennedy Park, East Birchmount Park   
7                     Oakridge, Clairlea, Golden Mile   
8     Scarborough Village West, Cliffcrest, Cliffside   
9                         Birch Cliff, Cliffside West   
10  Wexford Heights, Scarborough Town Centre, Dors...   
11                                  Maryvale, Wexford   
12                                          Agincourt   
13            Clarks Corners, Sullivan, Tam O'Shanter   
15                                    L'Amoreaux West   
16                                  Hillcrest Village   
17                       Fairview, Henry Farm, Oriole   
18                                    Bayview Village   
19                           Silver Hills, York Mills   
20                                   Willowdale South   
22                                    Willowdale West   
24                                    Don Mills North   
25                   Flemingdon Park, Don Mills South   
26    Wilson Heights, Bathurst Manor, Downsview North   
27                    York University, Northwood Park   
29                                     Downsview West   
30                                  Downsview Central   
31                                Downsview Northwest   
32                                   Victoria Village   
33                    Parkview Hill, Woodbine Gardens   
34                                   Woodbine Heights   
..                                                ...   
61                                           Roselawn   
63                North Midtown, Yorkville, The Annex   
64                     University of Toronto, Harbord   
65          Chinatown, Grange Park, Kensington Market   
66  Bathurst Quay, South Niagara, CN Tower, King a...   
67                    Stn A PO Boxes 25 The Esplanade   
68             Underground city, First Canadian Place   
69                   Lawrence Heights, Lawrence Manor   
71                                 Humewood-Cedarvale   
73                                           Christie   
74                       Dovercourt Village, Dufferin   
75                           Trinity, Little Portugal   
76       Brockton, Parkdale Village, Exhibition Place   
77                 Downsview, Upwood Park, North Park   
78     Del Ray, Mount Dennis, Keelesdale, Silverthorn   
79                      The Junction North, Runnymede   
80                      The Junction South, High Park   
81                             Parkdale, Roncesvalles   
82                                 Runnymede, Swansea   
83                                       Queen's Park   
84              Canada Post Gateway Processing Centre   
85  Business Reply Mail Processing Centre 969 Eastern   
86       New Toronto, Humber Bay Shores, Mimico South   
87                             Long Branch, Alderwood   
90  Mimico NW, Kingsway Park South West, Royal Yor...   
92  Old Burnhamthorpe, Markland Wood, Eringate, Bl...   
93                                      Humber Summit   
96                                          Westmount   
98  Albion Gardens, Thistletown, Jamestown, Humber...   
99                                          Northwest   

        1st Most Common Venue          2nd Most Common Venue  \
0        Fast Food Restaurant                     Print Shop   
1                         Bar     Construction & Landscaping   
2           Electronics Store                 Breakfast Spot   
3                 Coffee Shop              Korean Restaurant   
4            Hakka Restaurant                Thai Restaurant   
6            Department Store                    Coffee Shop   
7                    Bus Line                         Bake

Cluster 1

In [27]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
14  L'Amoreaux East, Agincourt North, Steeles East...                  Park   
21                                    York Mills West                  Park   
23                                          Parkwoods  Fast Food Restaurant   
28                        CFB Toronto, Downsview East                  Park   
38                                       East Toronto     Convenience Store   
42                                      Lawrence Park                  Park   
48                                           Rosedale                  Park   
62                Forest Hill North, Forest Hill West                  Park   
70                                          Glencairn                  Park   
72                                Caledonia-Fairbanks                  Park   
88      The Kingsway, Montgomery Road, Old Mill North                  Park   
95                                             Weston                  Park   
97  St. Phillips, Richview Gardens, Kingsview Vill...                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14      Asian Restaurant            Playground         Women's Store   
21                  Bank     Convenience Store                   Bar   
23                  Park     Food & Drink Shop         Women's Store   
28     Other Repair Shop               Airport         Women's Store   
38                  Park           Coffee Shop             Drugstore   
42              Bus Line           Swim School         Women's Store   
48            Playground              Building                 Trail   
62         Jewelry Store      Sushi Restaurant                 Trail   
70           Pizza Place   Japanese Restaurant                   Pub   
72         Women's Store              Pharmacy  Fast Food Restaurant   
88                  Pool                 River     Electronics Store   
95         Women's Store            Donut Shop          Dessert Shop   
97     Mobile Phone Shop           Pizza Place              Bus Line   

          5th Most Common Venue        6th Most Common Venue  \
14             Doner Restaurant                 Dessert Shop   
21                Women's Store                   Donut Shop   
23                   Donut Shop           Dim Sum Restaurant   
28                   Donut Shop                 Dessert Shop   
38           Dim Sum Restaurant                        Diner   
42             Doner Restaurant           Dim Sum Restaurant   
48  Eastern European Restaurant          Dumpling Restaurant   
62            Electronics Store  Eastern European Restaurant   
70                      Dog Run             Department Store   
72                       Market             Doner Restaurant   
88  Eastern European Restaurant          Dumpling Restaurant   
95           Dim Sum Restaurant                        Diner   
97                      Dog Run                 Dessert Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14    Dim Sum Restaurant                 Diner        Discount Store   
21                 Diner        Discount Store               Dog Run   
23                 Diner        Discount Store               Dog Run   
28    Dim Sum Restaurant                 Diner        Discount Store   
38        Discount Store               Dog Run      Doner Restaurant   
42                 Diner        Discount Store               Dog Run   
48     Electronics Store             Drugstore            Donut Shop   
62   Dumpling Restaurant             Drugstore            Donut Shop   
70          Dessert Shop    Dim Sum Restaurant                 Diner   
72          Dessert Shop    Dim Sum Restaurant                 Diner   
88             Drugstore            Donut Shop   Empanada Restaurant   
95        Discount Store               Dog Run      Doner Restaurant   
97    Dim Sum Restaurant                 Diner        Discount Store   

   10th Most Common

Cluster 2

In [28]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
91  Princess Gardens, Martin Grove, Islington, Clo...                  Bank   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
91         Women's Store            Donut Shop    Dim Sum Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
91                 Diner        Discount Store               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
91      Doner Restaurant             Drugstore       Department Store

Cluster 3

In [29]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
5           Scarborough Village            Playground         Women's Store   
46  Moore Park, Summerhill East          Tennis Court            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5             Donut Shop          Dessert Shop    Dim Sum Restaurant   
46         Women's Store      Doner Restaurant      Department Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5                  Diner        Discount Store               Dog Run   
46          Dessert Shop    Dim Sum Restaurant                 Diner   

   9th Most Common Venue 10th Most Common Venue  
5       Doner Restaurant              Drugstore  
46        Discount Store                Dog Run

Cluster 4

In [30]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
89  Kingsway Park South East, Humber Bay, The Quee...        Baseball Field   
94                                   Emery, Humberlea        Baseball Field   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
89             Drugstore    Dim Sum Restaurant                 Diner   
94             Drugstore    Dim Sum Restaurant                 Diner   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
89        Discount Store               Dog Run      Doner Restaurant   
94        Discount Store               Dog Run      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
89            Donut Shop         Women's Store       Department Store  
94            Donut Shop         Women's Store       Department Store

Let's check which neighborhoods belong to each cluster

In [31]:
for i in range(kclusters):
    print('Cluster ' + str(i) + ': \n {}'.format('\n '.join(toronto_merged[toronto_merged["Cluster Labels"] == i]["Neighborhood"])))

Cluster 0: 
 Rouge, Malvern
 Rouge Hill, Highland Creek, Port Union
 Morningside, West Hill, Guildwood
 Woburn
 Cedarbrae
 Ionview, Kennedy Park, East Birchmount Park
 Oakridge, Clairlea, Golden Mile
 Scarborough Village West, Cliffcrest, Cliffside
 Birch Cliff, Cliffside West
 Wexford Heights, Scarborough Town Centre, Dorset Park
 Maryvale, Wexford
 Agincourt
 Clarks Corners, Sullivan, Tam O'Shanter
 L'Amoreaux West
 Hillcrest Village
 Fairview, Henry Farm, Oriole
 Bayview Village
 Silver Hills, York Mills
 Willowdale South
 Willowdale West
 Don Mills North
 Flemingdon Park, Don Mills South
 Wilson Heights, Bathurst Manor, Downsview North
 York University, Northwood Park
 Downsview West
 Downsview Central
 Downsview Northwest
 Victoria Village
 Parkview Hill, Woodbine Gardens
 Woodbine Heights
 The Beaches
 Leaside
 Thorncliffe Park
 Riverdale, The Danforth West
 The Beaches West, India Bazaar
 Studio District
 Davisville North
 North Toronto West
 Davisville
 Rathnelly, Deer Park, So